pypeit/specobj.py

```diff
@@ -0,0 +1,492 @@
+        Only called if there *isn't* an attribute with this name
+        """
+        try:
+            return self.__getitem__(item)
+        except KeyError:
+            raise AttributeError(item)
+
+    def __setattr__(self, item, value):
+
+        if not '_SpecObj__initialised' in self.__dict__:  # this test allows attributes to be set in the __init__ method
+            return dict.__setattr__(self, item, value)
+        elif item in self.__dict__:       # any normal attributes are handled normally
+            dict.__setattr__(self, item, value)
+        else:
+            self.__setitem__(item, value)
+
+    def __setitem__(self, item, value):
+        if item not in data_model.keys():
+            raise IOError("Cannot set {} attribute.  It is not in the data model".format(item))
+        if not isinstance(value, data_model[item]['otype']):
+            print("Wrong data type for attribute: {}".format(item))
+            print("Allowed type(s) are: {}".format(data_model[item]['otype']))
+            raise IOError("Try again")
+        if isinstance(value, np.ndarray):
+            self._data[item] = value
+        else:
+            self._data.meta[item] = value
+
+    def __getitem__(self, item):
+        if item in self._data.keys():
+            return self._data[item].data
+        elif item in self._data.meta.keys():
+            return self._data.meta[item]
+        else:
+            raise KeyError
+
+    def set_name(self):
+        """
+        Generate a unique index for this spectrum based on the
+        slit/order, its position and for multi-slit the detector.
+
+        Sets self.name internally
+
+        Returns:
+            str: :attr:`self.name`
+
+        """
+        if 'Echelle' in self.PYPELINE:
+            # ObjID
+            self.name = naming_model['obj']
+            if 'ECH_FRACPOS' not in self._data.meta.keys():
+                self.name += '----'
+            else:
+                self.name += '{:04d}'.format(int(np.rint(1000*self.ECH_FRACPOS)))
+            # Order
+            self.name += '-'+naming_model['order']
+            self.name += '{:04d}'.format(self.ECH_ORDER)
+        else:
+            # Spat
+            self.name = naming_model['spat']
+            if 'SPAT_PIXPOS' not in self._data.meta.keys():
+                self.name += '----'
+            else:
+                self.name += '{:04d}'.format(int(np.rint(self.SPAT_PIXPOS)))
+            # Slit
+            self.name += '-'+naming_model['slit']
+            self.name += '{:04d}'.format(self.SLITID)
+        # Detector
+        sdet = parse.get_dnum(self.DET, prefix=False)
+        self.name += '-{:s}{:s}'.format(naming_model['det'], sdet)
+        # Return
+        return self.name
+
+
+    def copy(self):
+        """
+        Generate a copy of this object
+
+        Returns:
+            SpecObj
+
+        """
+        sobj_copy = SpecObj(self.PYPELINE, self.DET,
+                            indict=self.__dict__.copy())
```

Comment:
Is this the main use of `indict`?  Is it dangerous to have this visible to users?  Or is there no way around this?

Reply:
yes, it is the only use.  Not sure about dangerous..